// @ton/blueprint/dist/cli/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.helpMessages = exports.helpArgs = exports.templateTypes = void 0;
const chalk_1 = __importDefault(require("chalk"));
exports.templateTypes = [
    {
        name: 'An empty contract (FunC)',
        value: 'func-empty',
    },
    {
        name: 'An empty contract (Tolk)',
        value: 'tolk-empty',
    },
    {
        name: 'An empty contract (Tact)',
        value: 'tact-empty',
    },
    {
        name: 'A simple counter contract (FunC)',
        value: 'func-counter',
    },
    {
        name: 'A simple counter contract (Tolk)',
        value: 'tolk-counter',
    },
    {
        name: 'A simple counter contract (Tact)',
        value: 'tact-counter',
    },
];
exports.helpArgs = { '--help': Boolean };
const availableCommands = ['create', 'run', 'build', 'set', 'help', 'test', 'verify', 'convert'];
exports.helpMessages = {
    help: `${chalk_1.default.bold('Usage:')} blueprint ${chalk_1.default.cyan('help')} [${chalk_1.default.yellow('command')}]

Displays this message if no command is specified, or displays detailed help for the specified command.

Blueprint is generally invoked as follows:
  ${chalk_1.default.cyan('blueprint')} ${chalk_1.default.yellow('[command]')} ${chalk_1.default.gray('[command-args]')} ${chalk_1.default.gray('[flags]')}

${chalk_1.default.bold('List of available commands:')}
${availableCommands.map(c => `- ${chalk_1.default.green(c)}`).join('\n')}`,
    create: `${chalk_1.default.bold('Usage:')} blueprint ${chalk_1.default.cyan('create')} ${chalk_1.default.yellow('[contract name]')} ${chalk_1.default.gray('[flags]')}

Creates a new contract together with supporting files according to a template.

Contract name must be specified in PascalCase and may only include characters a-z, A-Z, 0-9. If not specified on the command line, it will be asked interactively.

${chalk_1.default.bold('Flags:')}
${chalk_1.default.cyan('--type')} <type> - specifies the template type to use when creating the contract. If not specified on the command line, it will be asked interactively.

${chalk_1.default.bold('List of available types:')}
${exports.templateTypes.map((t) => `${chalk_1.default.cyan(t.value)} - ${t.name}`).join('\n')}`,
    run: `${chalk_1.default.bold('Usage:')} blueprint ${chalk_1.default.cyan('run')} ${chalk_1.default.yellow('[script name]')} ${chalk_1.default.gray('[flags]')}

Runs a script from the scripts directory.

Script name is matched (ignoring case) to a file in the scripts directory. If not specified on the command line, the available scripts will be presented interactively.

${chalk_1.default.bold('Flags:')}
${chalk_1.default.cyan('--mainnet')}, ${chalk_1.default.cyan('--testnet')} - selects network
${chalk_1.default.cyan('--custom')} [api-endpoint] - use a custom API
${chalk_1.default.cyan('--custom-version')} - API version (v2, v4)
${chalk_1.default.cyan('--custom-key')} - API key (v2 only)
${chalk_1.default.cyan('--custom-type')} - network type (custom, mainnet, testnet)
${chalk_1.default.cyan('--tonconnect')}, ${chalk_1.default.cyan('--deeplink')}, ${chalk_1.default.cyan('--mnemonic')} - deployer options
${chalk_1.default.cyan('--tonscan')}, ${chalk_1.default.cyan('--tonviewer')}, ${chalk_1.default.cyan('--toncx')}, ${chalk_1.default.cyan('--dton')} - explorer (default: tonviewer)`,
    build: `${chalk_1.default.bold('Usage:')} blueprint ${chalk_1.default.cyan('build')} ${chalk_1.default.yellow('[contract name]')} ${chalk_1.default.gray('[flags]')}

Builds the specified contract according to the respective .compile.ts file. For Tact contracts, all generated files will be placed in the ${chalk_1.default.cyan('build/<contract name>')} folder.

${chalk_1.default.bold('Flags:')}
${chalk_1.default.cyan('--all')} - builds all available contracts.`,
    set: `${chalk_1.default.bold('Usage:')} blueprint ${chalk_1.default.cyan('set')} <${chalk_1.default.yellow('key')}> [${chalk_1.default.yellow('value')}]

${chalk_1.default.bold('Available keys:')}
- ${chalk_1.default.cyan('func')} - overrides @ton-community/func-js-bin version.`,
    test: `${chalk_1.default.bold('Usage:')} blueprint ${chalk_1.default.cyan('test')} [...args]

Runs ${chalk_1.default.green('npm test [...args]')}, which by default executes ${chalk_1.default.green('jest')}.`,
    verify: `${chalk_1.default.bold('Usage:')} blueprint ${chalk_1.default.cyan('verify')} ${chalk_1.default.yellow('[contract name]')} ${chalk_1.default.gray('[flags]')}

Verifies a deployed contract on ${chalk_1.default.underline('https://verifier.ton.org')}.

${chalk_1.default.bold('Flags:')}
${chalk_1.default.cyan('--mainnet')}, ${chalk_1.default.cyan('--testnet')} - selects network
${chalk_1.default.cyan('--custom')} [api-endpoint] - use custom API (requires --custom-type)
${chalk_1.default.cyan('--custom-version')} - API version (v2 default)
${chalk_1.default.cyan('--custom-key')} - API key (v2 only)
${chalk_1.default.cyan('--custom-type')} - network type (mainnet, testnet)`,
    convert: `${chalk_1.default.bold('Usage:')} blueprint ${chalk_1.default.cyan('convert')} ${chalk_1.default.yellow('[path to build script]')}

Attempts to convert a legacy bash build script to a Blueprint compile wrapper.`,
};
